### DISCUSSION TITLE
commit history per project rather than global

### DISCUSSION THREAD

**Discussion by aspiers** [magit/magit#4503] [Ideas]
I love that `git-commit-mode` lets me hit `M-p` and `M-n` to move through commit message history, but since I'm commonly committing to many repos at once, all these unrelated commit message histories get mixed up into a single history.  And since the history ring buffer seems to be fairly small, it's common to find that I can't retrieve the history I need (typically as a basis for writing a similar commit message).

So would it be possible to make the history ring buffer local to each repository?  From a glance at `git-commit.el`, it seems to use the emacs built-in library `log-edit.el` which has a global `log-edit-comment-ring`, but I wonder if something clever could be done to work around that.  For example, is it possible for magit to store a ring buffer per repo, and then wrap advice around the `log-edit` functions which make `log-edit-comment-ring` local to the commit message buffer as part of preparing the buffer, and then storing the new message back to the appropriate per-repo ring on commit?

**Comment by tarsius** [comment]
Done in f805f6a7720aeb894e9a58f0b36769ebdb53dd7c.

To enable in just one repository add this to `<worktree>/.dir-locals.el`:

```
((nil (git-commit-use-local-message-ring . t)))
```

**Reply by aspiers** [reply]
Awesome, thanks so much!!

**Reply by twmr** [reply]
I agree that the commit-message history-cycling feature, which I use on an almost daily-basis, is really nice. I OTOH rely on the history not-being repo-local, because I use almost the same commit messages for changes in multiple repos.

**Reply by aspiers** [reply]
Well the good news for you is that it's configurable and opt-in...

**Reply by twmr** [reply]
Yes I know, but I wanted to mention that there is a/are use-case/s where the global history is useful.
